Multi-domain web user download

orangejenny opened the pull request:
## Summary
https://dimagi-dev.atlassian.net/browse/USH-449

## Feature Flag
Enterprise Permissions

## Product Description
![Screen Shot 2021-04-24 at 6 07 22 PM](https://user-images.githubusercontent.com/1486591/115974053-fa810880-a527-11eb-88c5-f677bcfa6223.png)


## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

Included int his PR.

### QA Plan

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-2964)

### Safety story
Risk is fairly limited to this feature and the mobile users download.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

orangejenny committed:
Updated CommCareUserForm to handle either mobile or web users

orangejenny committed:
Added FilteredWebUserDownload view

orangejenny committed:
Added support for web users to count and download views

orangejenny committed:
Added spinner during count request

orangejenny committed:
Added multi-domain support to download

orangejenny committed:
Added invitations to web user count and fixed multi-domain logic

orangejenny committed:
Added filtering logic to invitations

orangejenny committed:
Added tests, made fixes

orangejenny committed:
Split get_users_by_filters into more named functions

orangejenny committed:
Split out count_invitations_by_filters

orangejenny committed:
Switched booleans for user_type in dbaccessors

orangejenny committed:
Switched from flags to user type in UserFilterForm

orangejenny committed:
Fixed error condition

orangejenny committed:
Switched from flags to user type in views

orangejenny committed:
Added role filtering to invitations

orangejenny committed:
Removed help text from domains input

orangejenny committed:
Consolidated feature flag logic

orangejenny committed:
Updated decorator to look at flag, not mirrored domain config

orangejenny committed:
Fixed decorator

biyeun committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into jls/cross-domain-web-user-download

# Conflicts:
#	corehq/apps/users/forms.py

orangejenny committed:
Added missing param to _get_es_query call

orangejenny committed:
Merge branch 'jls/cross-domain-web-user-download' of github.com:dimagi/commcare-hq into jls/cross-domain-web-user-download

snopoke committed:
Merge branch 'master' into jls/cross-domain-web-user-download

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -121,11 +125,8 @@
-    TWO_STAGE_USER_PROVISIONING,
-    PARALLEL_USER_IMPORTS
-)
+from corehq import toggles
```
just a preference

orangejenny committed:
Merge branch 'master' into jls/cross-domain-web-user-download

orangejenny committed:
Fixed task count to include multiple domains

orangejenny committed:
Added admin role option to filter

orangejenny committed:
Removed location filter from web download

orangejenny committed:
Merge branch 'master' into jls/cross-domain-web-user-download

orangejenny committed:
DRYed up start of dump_usernames

orangejenny committed:
Fixed clean_role_id

orangejenny committed:
Added invitations to task count

orangejenny committed:
Removed changes related to filtering for admins

mkangia commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -1212,14 +1213,11 @@ class FilteredUserDownload(BaseManageCommCareUserView):
         context = self.main_context
-        count_view = "count_commcare_users" if self.include_mobile_users else "count_web_users"
-        context.update({'form': form, 'count_users_url': reverse(count_view, args=[domain])})
+        context.update({'form': form, 'count_users_url': reverse(self.count_view, args=[domain])})
```
nice!

mkangia commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -1210,25 +1211,44 @@ def page_url(self):
+@method_decorator([require_can_use_filtered_user_download], name='dispatch')
+class FilteredCommCareUserDownload(FilteredUserDownload):
+    urlname = 'filter_and_download_commcare_users'
+    user_type = MOBILE_USER_TYPE
```
I am still reviewing with few commits to go, and i saw the change from "include_" to user types, and i am still thinking that classes would be better. There is still a lot of "ifs" in the code which isn't easy to follow and is making things less cleaner.

mkangia replied:
> I did switch from flags to types in this PR, but that was largely because my flags had been named include_* which was poor naming since it didn't make it clear the flags are mutually exclusive.

using "only_" would have solved that but i still think passing user types was better for other reasons.

mkangia commented on corehq/apps/users/forms.py:
```diff
@@ -1177,7 +1177,7 @@ def __init__(self, *args, **kwargs):
         self.couch_user = kwargs.pop('couch_user')
         self.user_type = kwargs.pop('user_type')
-        if not(self.user_type not in [MOBILE_USER_TYPE, WEB_USER_TYPE]):
+        if self.user_type not in [MOBILE_USER_TYPE, WEB_USER_TYPE]:
```
Thinking that user_types are passed internally within the code itself, so why is this check really necessary? When could the values be other than these?

orangejenny replied:
It's just defensive.

mkangia replied:
Cool.

I think it just makes someone reading the code think that there can be other user_type which we are not supporting here or something like that. Fine to keep it as an extra safety net.

mkangia replied:
nit: Though, it might be good to have it at the entry point(s) only? iirc, it's at many places, thinking if its not needed everywhere.

orangejenny replied:
I have mixed feelings about this - I like having less code, but also, why not be defensive everywhere? I'm going to leave this as-is for now.

mkangia replied:
ya, i was just thinking over-defensive, if it was.

mkangia commented on corehq/apps/users/forms.py:
```diff
@@ -1206,14 +1206,21 @@ def __init__(self, *args, **kwargs):
+                crispy.Field('location_id'),
+                crispy.Field('columns'),
+            ]
+        fields += [crispy.Field('domains')]
```
This was resolved in a later commit by having role and location always present. :+1:

mkangia commented on corehq/apps/users/forms.py:
```diff
@@ -1169,7 +1169,6 @@ class UserFilterForm(forms.Form):
         required=False,
         label=_('Project Spaces'),
         widget=forms.SelectMultiple(attrs={'class': 'hqwebapp-select2'}),
-        help_text=_('Add project spaces containing the desired users'),
```
Ya, that help text seemed a bit too much but was thinking if it was added for a good reason.
As a rephrase, I was thinking "Add project spaces to get users from" just to make it explicit.

mkangia commented on corehq/toggles.py:
```diff
@@ -1472,7 +1472,11 @@ def _commtrackify(domain_name, toggle_is_enabled):
+        Bulk user management features:
+        For mobile users, enables filtered download page, bulk deletion page, and bulk lookup page.
+        For web users, enables filtered download page.
+    """,
```
so we are extending this new view to this old feature flag as well?

So the form for them would show only one domain as the option always, unless they have mirrored domains?

orangejenny replied:
yep, exactly

mkangia commented on corehq/apps/users/dbaccessors.py:
```diff
@@ -94,14 +123,47 @@ def get_commcare_users_by_filters(domain, user_filters, count_only=False):
         return query.count()
     user_ids = query.scroll_ids()
-    return map(CommCareUser.wrap, iter_docs(CommCareUser.get_db(), user_ids))
+    return map(CouchUser.wrap_correctly, iter_docs(CommCareUser.get_db(), user_ids))
```
nit: Probably does not matter, but should this by `CouchUser.get_db()`?

orangejenny replied:
skipping this

mkangia replied:
Fine to skip but just correcting on the typo i made,

instead of 
```
map(CouchUser.wrap_correctly, iter_docs(CommCareUser.get_db(), user_ids))
```
should this be
```
map(CouchUser.wrap_correctly, iter_docs(CouchUser.get_db(), user_ids))
```
since its returning couch users now and not commcare users.

mkangia commented on corehq/apps/users/bulk_download.py:
```diff
@@ -129,7 +132,6 @@ def get_user_role_name(domain_membership):
 
 
 def make_web_user_dict(user, location_cache, domain):
-    user = CouchUser.wrap_correctly(user['doc'])
```
So this is not needed now because its already wrapped correctly before calling this method?

orangejenny replied:
yep, the `wrap_correctly` call at the end of `_get_users_by_filters` takes care of it

mkangia commented on corehq/apps/users/dbaccessors.py:
```diff
@@ -94,14 +123,47 @@ def get_commcare_users_by_filters(domain, user_filters, count_only=False):
+
+    Applies "search_string" filter to the invitations' emails. This does not
+    support ES search syntax, it's just a case-insensitive substring search.
+    Ignores location filter.
```
i think this last statement should be avoided/rephrased to "other filters ignored".

In fact just say about the filters, it simply considers. In case a new filter is added to user in future, this documentation will become invalid.

orangejenny replied:
Updated in 5377d650580

mkangia commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -1230,6 +1230,28 @@ def get(self, request, domain, *args, **kwargs):
+
+    @method_decorator(require_can_edit_commcare_users)
+    def get(self, request, domain, *args, **kwargs):
+        return super().get(request, domain, *args, **kwargs)
```
Nope. Its taken from the object itself later.

mkangia commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -1123,25 +1124,44 @@ def page_url(self):
 
 
+@method_decorator([require_can_use_filtered_user_download], name='dispatch')
+class FilteredCommCareUserDownload(FilteredUserDownload):
```
Since this PR is making it cleaner user types by using mobile and web, instead of using "commcare" for mobile, do we want to name this as `FilteredMobileUserDownload` ?

orangejenny replied:
I see what you're saying, but I'm going to call this out of scope. The code is a mix of "commcare user" and "mobile user", I'm not sure which one wins out.

mkangia commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -1301,47 +1321,86 @@ def _excel_data(self, rows):
     else:
         res = bulk_download_users_async.delay(domain, download.download_id, user_filters,
-                                              is_web_download, owner_id=request.couch_user.get_id)
+                                              (user_type == WEB_USER_TYPE), owner_id=request.couch_user.get_id)
```
nit: I personally would favor having `is_web_download = user_type == WEB_USER_TYPE` earlier and then passing it in argument, or just pass user_type and let the called method handle it.

orangejenny replied:
Let's agree to disagree on this one. I don't want to make a named variable earlier in the function since it's only used in this one place, and I'm also not keen to change the signature of `bulk_download_users_async` in this PR.

mkangia commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -1301,47 +1321,86 @@ def _excel_data(self, rows):
-        track_workflow(request.couch_user.username, 'Domain filter used for mobile download')
-    is_web_download = False
-    if form.cleaned_data['columns'] == CommCareUserFilterForm.USERNAMES_COLUMN_OPTION:
+        user_type = 'mobile' if user_type == MOBILE_USER_TYPE else 'web'
```
So,

the passed user_type is already either mobile/web? Not sure why this if is needed here.

orangejenny replied:
just defensive

orangejenny replied:
I replied to this from the whole PR view, without looking at the whole context. Removed `user_type` in 82781343579

mkangia reviewed (commented):
Took a pass by commit, so some comments might be redundant, but still left them for my own trails of thoughts.

Overall, nothing blocking. Mostly queries and suggestions.

The one thing that felt uncomfortable was the user type resulting in "if/else" blocks in the common user form class. Even though it was an improvement from the "include_" type arguments.

Just like making separate methods for counts and records for mobile and web users each makes things much cleaner, I think making individual inherited classes would help as well. 
There was already some steps done in this direction by assigning user_type and count_url to classes, but I think it can be applied for the form class as well.
Unless you have already considered it and it is not worth the effort.

Although, if you wish to leave code as it is now and not stretch the PR, that would be cool as well.

orangejenny committed:
Merge branch 'master' into jls/cross-domain-web-user-download

orangejenny committed:
Removed conditionals

Ultimately, update_task_state will no-op if passed None.

orangejenny commented on corehq/apps/users/bulk_download.py:
```diff
@@ -205,9 +201,9 @@ def parse_mobile_users(domain, user_filters, task=None, total_count=None):
             if task:
-                DownloadBase.set_progress(task, n + current_user_downloaded_count, total_count)
-        current_user_downloaded_count += n + 1
+                DownloadBase.set_progress(task, current_user_downloaded_count, total_count)
```
I looked closer at this and went in the opposite direction: we're inconsistent about checking if the task exists before calling `set_progress`, and this gets handled by [update_task_state](https://github.com/dimagi/commcare-hq/blob/e1a47c5b7819d724978ff99fd36fe8d4f66cd356/corehq/ex-submodules/soil/progress.py#L122-L125) which can handle None, so I removed some conditionals in 357a1c0133a

orangejenny commented on corehq/apps/users/bulk_download.py:
```diff
@@ -260,9 +261,21 @@ def parse_web_users(domain, task=None, total_count=None):
         user_headers.extend(json_to_headers(
             {'location_code': list(range(1, max_location_length + 1))}
         ))
+    if is_multi_domain_download:
```
No, though that makes me think `is_multi_domain_download` might be better with a different name. This really represents if there are any domains included that are not the domain that the user is currently logged into.

mkangia replied:
Ah i see, so we want to add domain if 
1. there are many
2. or if its not the same as the one user is logged into

Could possibly be named, "includes_other_domains"?

orangejenny replied:
Renamed to `is_cross_domain` in 74008cdb17d

orangejenny commented on corehq/apps/users/static/users/js/filtered_download.js:
```diff
@@ -28,7 +28,7 @@ hqDefine('users/js/filtered_download', [
             success: function (data) {
                 var count = data.count;
-                var template = count === 1 ? gettext("Download <%- count %> user") : gettext("Download <%- count %> users");
+                var template = count === 1 ? gettext("Download <%- count %> user") : gettext("Download <%- count %> records");
```
I changed it because the web users page includes both users and invitations.

orangejenny replied:
Although end users may still think of invitations as "invited users". I don't feel strongly about this, do you think it was better before?

mkangia replied:
I think i saw this before invitations was included or I missed registering invitations were being included.

Still, I am not sure users understand "records". 
Is there a harm in still saying users? after all an invitation would also represent a user only, no? and the end user does not we have a separate invitations model :)

orangejenny replied:
Yeah, I think you're right about the end user's mental model. Switched back to "users" in c2d650767b3

orangejenny commented:
@mkangia Thanks for review, I responded to a few things and will look into others further. I think you found some potential permissions issues, so thanks especially for that.

I'm leaning against making further changes regarding user type vs flags, or moving more code into classes, for the sake of scope. We have a lot of upcoming work in the user admin UIs, so there'll be plenty of opportunity to keep refactoring.

mkangia commented:
> I think you found some potential permissions issues

which one?

> I'm leaning against making further changes regarding user type vs flags, or moving more code into classes, for the sake of scope. We have a lot of upcoming work in the user admin UIs, so there'll be plenty of opportunity to keep refactoring.

Sounds good.

orangejenny committed:
Renamed is_multi_domain_download => is_cross_domain

orangejenny committed:
Updated button text

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -1430,10 +1431,15 @@ def count_users(request, domain, include_mobile_users=False, include_web_users=F
+        user_count += get_users_by_filters(current_domain, user_filters, include_mobile_users=include_mobile_users,
                                            include_web_users=include_web_users, count_only=True)
+        if include_web_users:
+            user_count += Invitation.by_domain(current_domain).count()
```
Invitations are purposefully part of the single-domain download, which was built for USH, and the multi-domain download should be consistent with the single-domain download.

Users will show up in the multi-domain download once for each domain they belong to - in code terms, this is really a domain membership download. The columns are about 50/50 user columns vs membership columns: name & email will be repeated for users who belong to multiple domains, but role, last access/login, and location are all domain-specific.

mkangia replied:
> Invitations are purposefully part of the single-domain download, which was built for USH

Okay. But its a part of the usual product now. I suspect it would be confusing for others but most of them won't have invites to cause that confusion, and also something product then decided to go with.

> the multi-domain download should be consistent with the single-domain download

That makes sense.

Just a curious follow up,
What happens if someone edits the invitation in the downloaded sheet and uploads it back?

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -1430,10 +1431,15 @@ def count_users(request, domain, include_mobile_users=False, include_web_users=F
         user_filters = form.cleaned_data
     else:
         return HttpResponseBadRequest("Invalid Request")
+
```
I think this is okay given that both of the calling functions (`count_commcare_users` and `count_web_users`) have permissions checks.

orangejenny committed:
Removed redundant local variable

orangejenny committed:
Updated superclasses of FilteredUserDownload classes

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -1210,15 +1212,13 @@ def page_url(self):
 
 
-@method_decorator([FILTERED_BULK_USER_DOWNLOAD.required_decorator()], name='dispatch')
 class FilteredUserDownload(BaseManageCommCareUserView):
```
Thank you, this also would have been a permissions bug. Updated in adc014127d2 so that permissions checks and breadcrumbs should be correct now.

orangejenny committed:
Updated Invitation.by_domain args

orangejenny commented on corehq/apps/users/models.py:
```diff
@@ -2676,8 +2676,10 @@ class Invitation(models.Model):
-        return Invitation.objects.filter(domain=domain, is_accepted=False)
+    def by_domain(cls, domain, **filters):
+        if 'is_accepted' not in filters:
+            filters['is_accepted'] = False
```
It isn't intending to change the functionality - do you think it is? The intent it just to add the option to filter by attributes other than `is_accepted`, while still only returning pending invitations when called without params. Updated in 50949d9f6c8 to make the signature hopefully clearer.

mkangia replied:
oh sorry, i missed that the method originally was returning pending invitations only.

I would expect a method doing this to look like
```
@classmethod
def pending_invitations(domain, **filters):
    Invitation.objects.filter(domain=domain, is_accepted=False, **filters)
```

Changes now looks good.

orangejenny committed:
Updated comment

orangejenny commented on corehq/apps/users/bulk_download.py:
```diff
@@ -238,15 +238,14 @@ def parse_web_users(domain, user_filters, task=None, total_count=None):
             if task:
                 DownloadBase.set_progress(task, progress, total_count)
-        for invite in Invitation.by_domain(current_domain):
+        for invite in get_invitations_by_filters(current_domain, user_filters):
```
This should already have been getting only pending invitations.

orangejenny commented on corehq/apps/es/tests/utils.py:
```diff
@@ -54,3 +57,12 @@ def es_test(test):
+def populate_user_index(user_docs):
+    es = get_es_new()
+    ensure_index_deleted(USER_INDEX)
+    initialize_index_and_mapping(es, USER_INDEX_INFO)
```
I see what you're saying, but I'm going to leave this alone for now. If the next dev to use this function wants to update it, that's fine.

orangejenny committed:
Switched UI to use knockout

orangejenny committed:
Added disabling of domain-specific filters when filtering across domains

mkangia commented on corehq/apps/users/forms.py:
```diff
@@ -1170,15 +1170,18 @@ class CommCareUserFilterForm(forms.Form):
         self.domain = kwargs.pop('domain')
         self.couch_user = kwargs.pop('couch_user')
-        super(CommCareUserFilterForm, self).__init__(*args, **kwargs)
+        self.include_mobile_users = kwargs.pop('include_mobile_users', False)
```
(comment redundant post further changes to use user_types)

mkangia commented on corehq/apps/users/bulk_download.py:
```diff
@@ -230,27 +228,29 @@ def parse_mobile_users(domain, user_filters, task=None, total_count=None):
+        for user in get_web_users_by_filters(current_domain, user_filters):
+            user_dict = make_web_user_dict(user, location_cache, current_domain)
+            user_dicts.append(user_dict)
+            max_location_length = max(max_location_length, len(user_dict["location_code"]))
```
just confirming if this would happen,
```
>>> len(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: object of type 'NoneType' has no len()
>>> len('')
0
```

orangejenny replied:
It shouldn't, `location_code` always gets populated: https://github.com/dimagi/commcare-hq/blob/123eedf5523878a71b0f8bebe7fe0cec2e2f3ef6/corehq/apps/users/bulk_download.py#L145

mkangia replied:
not saying you added this but, location_code has location_codes, that is great :)

mkangia reviewed (approved):
much thanks for addressing the review @orangejenny

orangejenny commented:
permissions issue => https://github.com/dimagi/commcare-hq/pull/29609/commits/adc014127d2f39f1d2864d01c2099a308b0629c8
(previously, the web user download didn't require permission to edit web users)

orangejenny committed:
Fixed up UI

- Switched from disabling to hiding inputs, because disabling leaves any previously-selected value visible
- Moved domain input to the top, otherwise the slide animation looks weird
- Updated domains filter to not show unless there are mutiple domains

orangejenny committed:
Updated invitation filtering to pass qualified id

orangejenny committed:
Removed unused imports

orangejenny commented:
@mkangia QA just finished for this one. Could you glance over the last few qabug-related commits?

mkangia commented on corehq/apps/users/forms.py:
```diff
@@ -1214,12 +1214,10 @@ def __init__(self, *args, **kwargs):
+        self.fields['domains'].choices = [('all_project_spaces', _('All Project Spaces'))] + \
+                                         [(self.domain, self.domain)] + \
+                                         [(domain, domain) for domain in
+                                          DomainPermissionsMirror.mirror_domains(self.domain)]
```
just to understand this better,

1. was `[(self.domain, self.domain)]` twice earlier in case there were mirror domains?
2. how does this impact the UI? it made sense earlier when we were only show options here if needed. Now they would see "All Project Spaces" and the domain each time.

orangejenny replied:
This does change the UI. Previously, the "Project Spaces" filter was always displayed, but if there weren't mirror domains, it would only ever contain the current domain. Now the filter is only displayed if there are mirror domains.

mkangia replied:
So we are populating choices always, but just showing it only when needed. That sounds okay.

orangejenny committed:
Merge branch 'master' into jls/cross-domain-web-user-download

mkangia commented:
:tada:
